### DISCUSSION TITLE
Does HelmRelease support force recreate resource?

### DISCUSSION THREAD

**Discussion by RammusXu** [fluxcd/flux2#2854] [Q&A] [answered]
When my helm chart deployment labels are changed, HelmRelease is broken.
And this does not work.

```yaml
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: rammus-vector
  namespace: logging-nonprod
spec:
  ...
  upgrade:
    force: true
```

I also find this [comment for Kustomize](https://github.com/fluxcd/flux2/discussions/2324#discussioncomment-2109244), so, I'm wondering if HelmRelease support recreate resource which is caused by `field is immutable`.

Full error message:
```
message: "Helm upgrade failed: cannot patch \"rammus-vector\" with kind Deployment: Deployment.apps \"rammus-vector\" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{\"app.kubernetes.io/instance\":\"rammus-vector\", \"app.kubernetes.io/managed-by\":\"Helm\", \"app.kubernetes.io/name\":\"vector\", \"app.kubernetes.io/version\":\"0.21.2\", \"helm.sh/chart\":\"vector-3.0.0\"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable\n\nLast Helm logs:\n\nPatch Deployment \"rammus-vector\" in namespace logging-nonprod\nerror updating the resource \"rammus-vector\":\n\t cannot patch \"rammus-vector\" with kind Deployment: Deployment.apps \"rammus-vector\" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{\"app.kubernetes.io/instance\":\"rammus-vector\", \"app.kubernetes.io/managed-by\":\"Helm\", \"app.kubernetes.io/name\":\"vector\", \"app.kubernetes.io/version\":\"0.21.2\", \"helm.sh/chart\":\"vector-3.0.0\"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable\nPatch HorizontalPodAutoscaler \"rammus-vector\" in namespace logging-nonprod\nLooks like there are no changes for CronJob \"rammus-vector-ead-cronjob\"\nwarning: Upgrade \"rammus-vector\" failed: cannot patch \"rammus-vector\" with kind Deployment: Deployment.apps \"rammus-vector\" is invalid: spec.selector: Invalid value: v1.LabelSelector{MatchLabels:map[string]string{\"app.kubernetes.io/instance\":\"rammus-vector\", \"app.kubernetes.io/managed-by\":\"Helm\", \"app.kubernetes.io/name\":\"vector\", \"app.kubernetes.io/version\":\"0.21.2\", \"helm.sh/chart\":\"vector-3.0.0\"}, MatchExpressions:[]v1.LabelSelectorRequirement(nil)}: field is immutable"

```

**Comment by masterkain** [comment]
depending on the release and what you might lose I'd just remove the deployment, push, waiting for the resources to clean up and activate the release again

this is common also for example when helm charts upgrade themselves or their dependencies to an incompatible version that requires redeployments

**Comment by kingdonb** [comment, marked as answer]
`spec.upgrade.force` should use the Replace strategy according to this doc:

```
	// Force forces resource updates through a replacement strategy.
	// +optional
	Force bool `json:"force,omitempty"`
```

~Were you able to get this working?~

It does, which means a different thing than `delete and recreate` – replace is patch, which still cannot update immutable fields. (differs from `kubectl replace --force` which actually does a delete and recreate when needed)

**Reply by RammusXu** [reply]
This is exactly my example like I wrote.

**Reply by kingdonb** [reply]
I believe I found what's wrong. It is a Helm issue, working as intended:

https://github.com/helm/helm/issues/9433

The `helm upgrade --force` command no longer does a delete and recreate (this is what Kustomize controller's `spec.force` does, for example, when a resource has immutable fields and force is enabled, it gets deleted and recreated a-la `kubectl replace --force`)

Helm does not do that, it is not clear why this change was made other than that it's potentially dangerous to delete and recreate resources, which by itself is certainly a good reason, although I think it's a mistake to prevent users from accessing the capability entirely. It's really not clear to me then what force is used for anymore, but it looks like this decision stems from an even further upstream decision to do the same with `kubectl apply --force`.

https://github.com/kubernetes/kubernetes/issues/71042#issuecomment-441307539

It might not have been a mistake for `kubectl` where it is a mistake for `helm` to do this, because users of `kubectl` can choose to use `kubectl replace --force` but users of Helm are apparently not presented with any similar alternative path to access the foot-gun. But it is the decision Helm made, (and it is pretty clear that it was decided this way because it's understood to be a foot-gun.) It might make sense for Flux to reverse this and provide a `forceReplace` strategy that Helm upstream does not, but I cannot speak to the logistical possibility itself or about whether Flux actually could reverse it independently (without actually forking the Helm package upstream.)

**Reply by mnaser** [reply]
@kingdonb a few years later, but have you found some sort of workaround that is GitOps-y for this issue that doesn't involve removing resources, wait for reconcile, create a new one?

**Reply by kingdonb** [reply]
I'm still deleting resources manually, but we have `flux reconcile helmrelease --force` now - in case you haven't enabled drift detection, don't want to wait for the next interval to arrive, nor are able to push a trivial/no-op change to your helm values, this one is new since Flux 2.3 (I think? Which one made Helm Controller GA)

I still don't think Helm Controller should be deleting and recreating resources in any case, but if you have a specific use case that isn't served well, I'd be happy to talk about it with you in a more interactive setting (join us on the #flux channel in the CNCF slack!)

**Reply by Nicola-Sergio** [reply]
@kingdonb In case I was managing a statefulset via FluxCD to comply with the GitOps approach, and I modified non-modifiable parameters, what would be the reasoning behind not allowing to destroy that resource and recreate it to avoid the update error? For visibility of what happens or something else? I ask to better understand the issue

**Reply by stefanprodan** [reply]
Flux helm-controller is limited to Helm SDK features, if the Helm CLI does not recreate statefulsets when run with `helm upgrade --force` neither will Flux.

**Reply by mnaser** [reply]
That's fair enough, I get the idea of sticking to what the SDK provides in terms of features.  It's unfortunate that upstream seems to see it as "okay" to just "delete" resources manually, so I think that's something we'll have to push to fix in Helm if that's the case.

**Reply by Nicola-Sergio** [reply]
So what I would do when I have to modify unmodifiable parameter of whatever kind of deploy?
I usually do this:
- Suspend ks flux-system
- Delete manually the helm release 
- Push changes to the repo
- Ensure that the resource is completely deleted
- Resume ks and reconcile it with -- with-source flag

Is it the correct approach?
